samples/nrf9160/serial_lte_modem/src/slm_at_icmp.c

```diff
@@ -0,0 +1,516 @@
+				hcs = check_ics(data, len - header_len);
+				if (hcs != 0) {
+					LOG_ERR("HCS error %d", hcs);
+					goto close_end;
+				}
+				/* Payload length */
+				pllen = (ip_buf[2] << 8) + ip_buf[3];
+			}
+
+			/* Check seqnr and length */
+			plseqnr = data[7];
+			if (plseqnr != seqnr) {
+				LOG_ERR("error sequence numbers %d %d",
+					plseqnr, seqnr);
+				goto close_end;
+			}
+			if (pllen != len) {
+				LOG_ERR("error length %d %d", pllen, len);
+				goto close_end;
+			}
+
+			/* FIXME: Check Code */
+
+			/* Result */
+			sprintf(buf, "#XPING: %d.%03d\r\n",
+				(u32_t)(delta_t)/1000,
+				(u32_t)(delta_t)%1000);
+			m_callback(buf);
+		}
+
+close_end:
+		(void)nrf_close(fd);
+	} while (0);
+
+	return (u32_t)delta_t;
+}
+
+static void ping_task(void *arg1, void *arg2, void *arg3)
+{
+	struct addrinfo *si = argv.src;
+	struct addrinfo *di = argv.dest;
+	u32_t sum = 0;
+	u32_t count = 0;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	for (int i = 0; i < argv.count; i++) {
+		u32_t to = send_ping_wait_reply();
+
+		if (to > 0)  {
+			count++;
+			sum += to;
+		}
+		k_sleep(argv.interval);
+	}
+
+	if (count > 1) {
+		u32_t ave = (sum + count/2) / count;
+		int ave_s = ave / 1000;
+		int ave_f = ave % 1000;
+
+		sprintf(buf, "#XPING: average %d.%03d\r\n", ave_s, ave_f);
+		m_callback(buf);
+		m_callback("OK\r\n");
+	}
+
+	freeaddrinfo(si);
+	freeaddrinfo(di);
+	icmp_thread_id = NULL;
+}
+
+static int ping_test_handler(const char *url, int length, int waittime,
+			int count, int interval)
+{
+	int st;
+	struct addrinfo *res;
+	int addr_len;
+
+	st = modem_info_params_get(&modem_param);
+	if (st < 0) {
+		LOG_ERR("Unable to obtain modem parameters (%d)", st);
+		return -1;
+	}
+
+	/* Check network connection status by checking local IP address */
+	addr_len = strlen(modem_param.network.ip_address.value_string);
+	if (addr_len == 0) {
+		LOG_ERR("LTE not connected yet");
+		return -1;
+	}
+	if (!check_for_ipv4(modem_param.network.ip_address.value_string,
+			addr_len)) {
+		LOG_ERR("Invalid local address");
+		return -1;
+	}
+
+	/* Get own IP address */
+	st = getaddrinfo(modem_param.network.ip_address.value_string,
+			NULL, NULL, &res);
+	if (st != 0) {
+		LOG_ERR("getaddrinfo(src) error: %d", st);
+		freeaddrinfo(argv.dest);
+		return -st;
+	}
+	argv.src = res;
+
+	/* Get destination */
+	res = NULL;
+	st = getaddrinfo(url, NULL, NULL, &res);
+	if (st != 0) {
+		LOG_ERR("getaddrinfo(dest) error: %d", st);
+		return -st;
+	}
+	argv.dest = res;
+
+	if (argv.src->ai_family != argv.dest->ai_family) {
+		LOG_ERR("Source/Destination address family error");
+		freeaddrinfo(argv.dest);
+		freeaddrinfo(argv.src);
+		return -1;
+	}
+
+	argv.len = length;
+	argv.waitms = waittime;
+	argv.count = 1;		/* default 1 */
+	argv.interval = 1000;	/* default 1s */
+	if (count > 0) {
+		argv.count = count;
+	}
+	if (interval > 0) {
+		argv.interval = count;
+	}
+
+	icmp_thread_id = k_thread_create(&ping_thread, ping_thread_stack,
```

Comment:
@lemrey @SebastianBoe do you know perhaps if it's fine to spawn the same thread multiple times in the application runtime? Assuming of course that it has finished it's execution before. I've never seen such an use with Zephyr.

Probably a safer approach would be to create a custom workqueue and submit a workqueue item. This should not have much overhead over a simple thread. But let's hear from others first.

Reply:
Not an expert on runtime, but I doubt it is OK.

Reply:
The original design is to allow the thread to quit naturally after the rounds of ping have been done. Since it involves blocking socket API call, a thread might be better than Works, which is in a global work queue.

Reply:
What I suggested here was not to use a (global) system workqueue, but a custom one: https://docs.zephyrproject.org/1.9.0/kernel/threads/workqueues.html#defining-a-workqueue.

There's really no benefit in closing a thread, as we are not able to re-use its resources anyway when it's not running. And given the fact, it's uncertain if restarting a thread is supported, I still think we should use a workqueue.

Reply:
I tired the local work queue and it works. NOTE the sample code in the URL still submit to system queue, and I changed to submit to my own queue. Actually there could be an issue in my next PR where enter/exit Idle mode is implemented. I might change to use system queue if user queue cannot survive the process.